eureka-write-server/src/main/java/com/netflix/eureka2/server/channel/RegistrationChannelImpl.java

```diff
@@ -77,97 +112,145 @@ public void onNext(Void aVoid) {
+                    }
+                });
             }
         }
 
-        final long tempNewVersion = currentVersion + 1;
-        final InstanceInfo tempNewInfo = new InstanceInfo.Builder()
-                .withInstanceInfo(instanceInfo).withVersion(tempNewVersion).build();
+        logger.debug("Registering service in registry: {}", instanceInfo);
 
-        Observable<Boolean> registerResult = registry.register(tempNewInfo);
-        registerResult.subscribe(new Subscriber<Boolean>() {
-            @Override
-            public void onCompleted() {
-                currentVersion = tempNewVersion;
-                currentInfo = tempNewInfo;
-                sendAckOnTransport();
-            }
+        final long tempNewVersion = currentVersion.getAndIncrement();
 
-            @Override
-            public void onError(Throwable e) {
-                sendErrorOnTransport(e);
-            }
+        if (cachedInfo.get() == null) {  // setup initial cache
+            cachedInfo.set(new InstanceInfo.Builder().withInstanceInfo(instanceInfo).withVersion(tempNewVersion - 1).build());
+        }
 
-            @Override
-            public void onNext(Boolean status) {
-                // No op
-            }
-        }); // Callers aren't required to subscribe, so it is eagerly subscribed.
-        return registerResult.ignoreElements().cast(Void.class);
+        final InstanceInfo tempNewInfo = new InstanceInfo.Builder()
+                .withInstanceInfo(instanceInfo).withVersion(tempNewVersion).build();
+
+        return registry.register(tempNewInfo)
+                .ignoreElements()
+                .cast(Void.class)
+                .doOnError(new Action1<Throwable>() {
+                    @Override
+                    public void call(Throwable throwable) {
+                        sendErrorOnTransport(throwable).doOnTerminate(new Action0() {
+                            @Override
+                            public void call() {
+                                close();
+                            }
+                        });
+                    }
+                })
+                .doOnCompleted(new Action0() {
+                    @Override
+                    public void call() {
+                        sendAckOnTransport().doOnError(new Action1<Throwable>() {
+                            @Override
+                            public void call(Throwable throwable) {
+                                logger.warn("Failed to send ack for register operation for instanceInfo {}", cachedInfo.get());
+                            }
+                        });
+                    }
+                });
     }
 
+    /**
+     * Cases:
+     * 1. channel state is Registered. Call unregister on the registry, and
+     *   1a. if successful, ack on the channel and close the channel regardless of the ack result
+     *   1b. if unsuccessful, send error on the transport. TODO should we optimize and close the channel here?
```

Comment:
There is no concept of sending an error on the transport. I think we should close eagerly here.